src/recordlinker/schemas/algorithm.py

```diff
@@ -44,6 +44,86 @@ def serialize_func(self, value: matchers.FeatureFunc) -> str:
         return str(value)
 
 
+class LogOdd(pydantic.BaseModel):
+    """
+    The schema for an LogOdd record.
+    """
+
+    model_config = pydantic.ConfigDict(from_attributes=True)
+
+    feature: Feature = pydantic.Field(json_schema_extra={"enum": Feature.all_options()})
+    value: Annotated[float, pydantic.Field(ge=0)]
+
+    @pydantic.field_validator("feature", mode="before")
+    def validate_feature(cls, value):
+        """
+        Validate the feature is a valid PII feature.
+        """
+        try:
+            return Feature.parse(value)
+        except ValueError as e:
+            raise ValueError(f"Invalid feature: '{value}'. {e}")
+
+
+class SkipValue(pydantic.BaseModel):
+    feature: str = pydantic.Field(json_schema_extra={"enum": Feature.all_options() + ["*"]})
+    values: list[str] = pydantic.Field(min_length=1)
+
+    @pydantic.field_validator("feature", mode="before")
+    def validate_feature(cls, value):
+        """
+        Validate the feature is a valid PII feature.
+        """
+        if value == "*":
+            return value
+        try:
+            Feature.parse(value)
+        except ValueError as e:
+            raise ValueError(f"Invalid feature: '{value}'. {e}")
+        return value
+
+
+class AlgorithmContext(pydantic.BaseModel):
+    """
+    The schema for an algorithm context record.
+    """
+
+    model_config = pydantic.ConfigDict(from_attributes=True)
+
+    include_multiple_matches: bool = True
+    log_odds: typing.Sequence[LogOdd] = []
+    skip_values: typing.Sequence[SkipValue] = []
+
+    @pydantic.model_validator(mode="after")
+    def init_log_odds_helpers(self) -> typing.Self:
+        """
+        Initialize cache helpers for returning log odds values.
+        """
+        self._log_odds_cache: dict[str, float | None] = {}
+        self._log_odds_mapping: dict[str, float] = {str(o.feature): o.value for o in self.log_odds}
```

Comment:
Is this right here basically the answer to my question earlier? Under the hood it really is a dictionary that's being accessed? I think I'm following, I just want to make sure I understand how this part of the refactor actually works because it's not something I've used before.

Reply:
Yep, I'm using the feature/value attributes to give the schema structure that's easy to validate and document.  The `get_log_odds` method makes it quick (via caching) and easy to programmatically retrieve values.